### ISSUE TITLE
cyclus-build-deps in need of update

### ISSUE THREAD

**Issue by dermesser** [cyclus/cyclus#1562]
Also relates to #1497.

I have some hard-earned knowledge on building cyclus within an anaconda environment. While initially being led astray by `cyclus-build-deps`, and the outdated INSTALL and DEPENDENCIES files, I have managed to assemble an anaconda environment that enables building cyclus in a reliable manner.

As Anaconda YAML file, the following works (for me), for cyclus 1.5.5; please let me know if you'd be interested in adding such information to the existing install documentation. Especially fickle details, like which boost package to use, linking errors among glib/xml++/..., are time-consuming to figure out.

```yaml
channels:
  - conda-forge
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - _openmp_mutex=4.5=1_gnu
  - _sysroot_linux-64_curr_repodata_hack=3=haa98f57_10
  - binutils_impl_linux-64=2.35.1=h27ae35d_9
  - binutils_linux-64=2.35.1=h454624a_30
  - blas=1.1=openblas
  - boost-cpp=1.70.0=h7b93d67_3
  - bzip2=1.0.8=h7f98852_4
  - ca-certificates=2021.5.25=h06a4308_1
  - certifi=2021.5.30=py36h06a4308_0
  - cmake=3.19.6=h973ab73_0
  - coincbc=2.9.9=blas_openblash0f40ac4_1002
  - cython=0.29.23=py36h2531618_0
  - expat=2.4.1=h2531618_2
  - gcc_impl_linux-64=9.3.0=h6df7d76_17
  - gcc_linux-64=9.3.0=h1ee779e_30
  - gettext=0.19.8.1=h0b5b191_1005
  - glib=2.68.3=h9c3ff4c_0
  - glib-tools=2.68.3=h9c3ff4c_0
  - glibmm=2.58.1=h778ead0_1
  - gxx_impl_linux-64=9.3.0=hbdd7822_17
  - gxx_linux-64=9.3.0=h7e70986_30
  - hdf5=1.10.6=hb1b8bf9_0
  - icu=67.1=he1b5a44_0
  - jinja2=3.0.0=pyhd3eb1b0_0
  - kernel-headers_linux-64=3.10.0=h57e8cba_10
  - krb5=1.18.2=h173b8e3_0
  - ld_impl_linux-64=2.35.1=h7274673_9
  - libblas=3.8.0=17_openblas
  - libcblas=3.8.0=17_openblas
  - libcurl=7.71.1=h20c2e04_1
  - libedit=3.1.20210216=h27cfd23_1
  - libffi=3.3=he6710b0_2
  - libgcc-devel_linux-64=9.3.0=hb95220a_17
  - libgcc-ng=9.3.0=h5101ec6_17
  - libgfortran-ng=7.5.0=h14aa051_19
  - libgfortran4=7.5.0=h14aa051_19
  - libglib=2.68.3=h3e27bee_0
  - libgomp=9.3.0=h5101ec6_17
  - libiconv=1.16=h516909a_0
  - liblapack=3.8.0=17_openblas
  - libopenblas=0.3.10=h5a2b251_0
  - libsigcpp=2.10.3=ha770c72_0
  - libssh2=1.9.0=h1ba5d50_1
  - libstdcxx-devel_linux-64=9.3.0=hf0c5c8d_17
  - libstdcxx-ng=9.3.0=hd4cf53a_17
  - libuv=1.40.0=h7b6447c_0
  - libxml2=2.9.10=h68273f3_2
  - libxmlpp=2.40.1=h0bd56ce_1004
  - lz4-c=1.9.3=h9c3ff4c_0
  - markupsafe=2.0.1=py36h27cfd23_0
  - ncurses=6.2=he6710b0_1
  - numpy=1.19.5=py36h2aa4a07_1
  - openblas=0.3.3=h9ac9557_1001
  - openssl=1.1.1k=h27cfd23_0
  - pandas=1.1.5=py36ha9443f7_0
  - pcre=8.45=h9c3ff4c_0
  - pip=21.1.2=py36h06a4308_0
  - pkg-config=0.29.2=h1bed415_8
  - python=3.6.13=h12debd9_1
  - python-dateutil=2.8.1=pyhd3eb1b0_0
  - python_abi=3.6=1_cp36m
  - pytz=2021.1=pyhd3eb1b0_0
  - readline=8.1=h27cfd23_0
  - rhash=1.4.1=h3c74f83_1
  - setuptools=52.0.0=py36h06a4308_0
  - sigcpp-2.0=2.10.3=h58526e2_0
  - six=1.16.0=pyhd3eb1b0_0
  - sqlite=3.35.4=hdfb4753_0
  - sysroot_linux-64=2.17=h57e8cba_10
  - tk=8.6.10=hbc83047_0
  - wheel=0.36.2=pyhd3eb1b0_0
  - xz=5.2.5=h7b6447c_0
  - zlib=1.2.11=h7b6447c_3
  - zstd=1.4.9=ha95c52a_0
```

**Comment by gonuke** [comment]
Thanks for all the effort to sort this out.  We're always happy to take input like this and update our installation documentation, but I wonder if there are deeper lessons about how we can make the installation more straightforward?  Do you think there are other changes we could make that would render some of this less important/fragile?

Hopefully, we'll soon have some resources to review and update much of our build/install infrastructure.

**Comment by dermesser** [comment]
I have to add a caveat: With these dependencies, cyclus builds and runs, but I've experienced spurious crashes (e.g. in the python library). The docker image works well for me, and I'm still trying other ways to build cyclus directly. About that I'll keep this issue posted though :)

FWIW, I don't seem to be alone. See #1273, from 2016: *After more all-consuming weeks of my life wasted #notbitter, I finally have a version of Cyclus that is building & testing on conda-forge on Linux and Mac. See the end of conda-forge/staged-recipes#678 for proof.* ;-)

>  Do you think there are other changes we could make that would render some of this less important/fragile?

(rambling ahead, I'm sorry)

A good question, which I have asked myself too, because attempting to build cyclus has me question the way that things in software are done today, on a very general level. While I acknowledge that it's mostly impossible to make such a complex piece of software build on any random Linux distribution due to the version hell of shared libraries, it would be very nice to not only have to rely on Docker or Virtualbox images. Thus, from my perspective, providing the information to enable building Cyclus from source within an anaconda environment would be both enough and very useful. It used to be (apparently? I am guessing from what I can see) possible to use `cyclus-build-deps` to do exactly that, and having that up-to-date would be a good step forward.

The other part is that due to the great flexibility of cyclus, the build system seems quite complex. So while it works well if it works, in the past weeks I've struggled hard with library issues that were really difficult to track down, and sometimes vanished on their own. Just as an example; my current favorite is this error below, despite having libxml2 and libiconv installed, found by pkg-config, etc. (this happens with the dependencies used by the official docker image, on my own machine):

```
/home/lbo/.anaconda3/envs/fuelcycle_docker/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: cannot find -lxml2
/home/lbo/.anaconda3/envs/fuelcycle_docker/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: cannot find -liconv
```

My past weeks have been filled with issues such as this one.

How to solve this in general? Again, I don't think there is an easy way. A lot of time and effort has gone into the current build system, enabling it to work in many places. But at the same time, it doesn't really anymore, right? At least not in a current Linux distribution, nor in anaconda environments as far as I have tried. The INSTALL and DEPENDENCIES documentation also is not helping anymore. The only way I've reliably managed to build Cyclus is by using the provided Dockerfile (respectively the docker image), and building cyclus inside of that. Maybe that's just what should be the accepted approach, and cyclus is to document this situation. In addition, documentation surrounding docker (or singularity, for the hard-core HPC scientists), e.g. how to mount your own source code into a container, may help alleviate the lack of a convenient build-on-your-own-machine option.

Apologies for the ranting, and I want to emphasize that I do not see any fault with cyclus. It's just the world we have to live in, as programmers,, and I'm looking forward to finding ways to improve the current situation :))

**Comment by gonuke** [comment]
Rant accepted! We have our own rants on similar topics - it is challenging to maintain complex software with a small team.

**Comment by dermesser** [comment]
A more humble goal for now would probably be updating the website and repository's documentation about installing Cyclus from source: it makes it seem very easy, but points readers the wrong way, as the shown required dependencies (e.g. at https://fuelcycle.org/user/install_from_git.html) are not present in anaconda anymore.

As so often, wrong information is worse than no information, so just by removing the outdated instructions and saying "if you want to run Cyclus, try the docker or virtualbox images" (or similar, whatever you think is best), this would already help beginners like me have an easier start :)

**Comment by gonuke** [comment]
Would you like to submit a PR to this effect? No pressure, but thought I'd check before seeing who else might be better suited to tackling this.

**Comment by gonuke** [comment]
I think this is resolved by #1588.  Closing, but please reopen if issues persist.
